wfe/wfe.go

```diff
@@ -938,28 +940,66 @@ func (wfe *WebFrontEndImpl) Challenge(
 		wfe.sendError(response, logEvent, probs.NotFound("No such challenge"), nil)
 	}
 
-	// Challenge URIs are of the form /acme/challenge/<auth id>/<challenge id>.
-	// Here we parse out the id components.
+	// Challenge URIs are of the form /acme/challenge/<auth id>/<challenge id>
+	// or /acme/challenge/v2/<auth id>/<challenge id> depending on the authorization
+	// version. Here we parse out the authorization and challenge IDs and retrieve
+	// the authorization.
 	slug := strings.Split(request.URL.Path, "/")
-	if len(slug) != 2 {
+	if len(slug) != 2 && len(slug) != 3 {
```

Comment:
Rather than having differing lengths of slugs, what do you think of having a different handler for v2 challenges? And do this at the HandleFunc level. I.e. `wfe.HandleFunc(m, "/acme/challengev2/", wfe.ChallengeV2)`. Then `Challenge` and `ChallengeV2` could each handle the stuff that is specific to them, with any commonalities factored into a third function they both call.

The same structure could apply to Authorizations.

Reply:
I'm not sure I agree this makes things significantly cleaner, the only thing that would get factored out as different between the two handlers in both cases is retrieving the v2 authorization and any work done to move all common code into a third function would be undone once we move to only supporting the v2 form, basically brings us right back to what we have now pre-change. I agree this isn't the cleanest looking code at the moment but I think it sets us up for an easier removal down the road.

Reply:
Alright, sounds good.